server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java

```diff
@@ -322,19 +312,245 @@ private void finishBloomFilterStoredFormat() throws IOException {
             IOUtils.close(toClose);
         }
 
         @Override
         public long ramBytesUsed() {
-            return buffer.ramBytesUsed() + delegateWriter.ramBytesUsed();
+            return bloomFilterWriter == null ? 0 : bloomFilterWriter.buffer.ramBytesUsed() + delegateWriter.ramBytesUsed();
+        }
+
+        private void maybeInitializeBloomFilterWriter(FieldInfo fieldInfo, int bitSetSizeInBits) throws IOException {
+            assert isPowerOfTwo(bitSetSizeInBits) : "Expected a power of two but got " + bitSetSizeInBits;
+            if (bloomFilterWriter != null) {
+                return;
+            }
+
+            try {
+                bloomFilterWriter = new BloomFilterWriter(fieldInfo, bitSetSizeInBits);
+                toClose.add(bloomFilterWriter);
+            } catch (IOException e) {
+                IOUtils.closeWhileHandlingException(toClose);
+                throw e;
+            }
+        }
+
+        class BloomFilterWriter implements Closeable {
+            private final FieldInfo fieldInfo;
+            private final int bitsetSizeInBits;
+            private final int bitSetSizeInBytes;
+            private final ByteArray buffer;
+            private final int[] hashes;
+            private final IndexOutput bloomFilterDataOut;
+
+            private boolean flushed = false;
+
+            BloomFilterWriter(FieldInfo fieldInfo, int bitsetSizeInBits) throws IOException {
+                this.fieldInfo = fieldInfo;
+                this.bitsetSizeInBits = bitsetSizeInBits;
+                this.bitSetSizeInBytes = bitsetSizeInBits / Byte.SIZE;
+                this.buffer = bigArrays.newByteArray(bitSetSizeInBytes, false);
+                this.hashes = new int[numHashFunctions];
+                this.bloomFilterDataOut = directory.createOutput(bloomFilterFileName(segmentInfo, segmentSuffix), context);
+
+                CodecUtil.writeIndexHeader(
+                    bloomFilterDataOut,
+                    STORED_FIELDS_BLOOM_FILTER_FORMAT_NAME,
+                    VERSION_CURRENT,
+                    segmentInfo.getId(),
+                    segmentSuffix
+                );
+            }
+
+            private void addToBloomFilter(FieldInfo info, BytesRef value) {
+                ensureNotFlushed();
+
+                assert info.getName().equals(bloomFilterFieldName) : "Expected " + bloomFilterFieldName + " but got " + info;
+                var termHashes = hashTerm(value, hashes);
+                for (int hash : termHashes) {
+                    final int posInBitArray = hash & (bitsetSizeInBits - 1);
+                    final int pos = posInBitArray >> 3; // div 8
+                    final int mask = 1 << (posInBitArray & 7); // mod 8
+                    final byte val = (byte) (buffer.get(pos) | mask);
+                    buffer.set(pos, val);
+                }
+            }
+
+            private void mergeBloomFiltersWithOr(MergeState mergeState) throws IOException {
```

Comment:
```suggestion
            private void mergeOptimized(MergeState mergeState) throws IOException {
```

Reply:
Done in https://github.com/elastic/elasticsearch/pull/137622/commits/07ad165b4088ad478873867476c8fdc00cf2777a